prov/verbs/src/verbs_info.c

```diff
@@ -885,29 +886,45 @@ int fi_ibv_get_srcaddr_devs(struct fi_info *info)
 		return 0;
 	}
 
-	for (fi = info; fi; fi = fi->next) {
+	for (fi = *info; fi;) {
+		dev_found = 0;
 		dlist_foreach_container(&verbs_devs, dev, entry)
-			if (!strncmp(fi->domain_attr->name, dev->name, strlen(dev->name)))
+			if (!strncmp(fi->domain_attr->name, dev->name, strlen(dev->name))) {
+				dev_found = 1;
 				break;
+			}
 
-		dlist_foreach_container(&dev->addrs, addr, entry) {
-			/* When a device has multiple interfaces/addresses configured
-			 * duplicate fi_info and add the address info. fi->src_addr
-			 * would have been set in the previous iteration */
-			if (fi->src_addr) {
-				if (!(add_info = fi_dupinfo(fi))) {
-					ret = -FI_ENOMEM;
-					goto out;
+		if(dev_found) {
+			dlist_foreach_container(&dev->addrs, addr, entry) {
+				/* When a device has multiple interfaces/addresses configured
+				 * duplicate fi_info and add the address info. fi->src_addr
+				 * would have been set in the previous iteration */
+				if (fi->src_addr) {
+					if (!(add_info = fi_dupinfo(fi))) {
+						ret = -FI_ENOMEM;
+						goto out;
+					}
+
+					add_info->next = fi->next;
+					fi->next = add_info;
+					fi = add_info;
 				}
 
-				add_info->next = fi->next;
-				fi->next = add_info;
-				fi = add_info;
+				ret = fi_ibv_rai_to_fi(addr->rai, fi);
+				if (ret)
+					goto out;
 			}
-
-			ret = fi_ibv_rai_to_fi(addr->rai, fi);
-			if (ret)
-				goto out;
+			fi = fi->next;
+		} else { /* src_addr required to bind listener. exclude info
+			    without src_addr.
+			    this situation may happen when interface is up,
+			    but not configured (no IP address) */
+			if(fi == *info)
+				*info = fi->next;
+			to_remove = fi;
+			fi = fi->next;
+			to_remove->next = NULL;
+			fi_freeinfo(to_remove);
```

Comment:
I guess this would be a problem when there are multiple IPoIB interfaces with only some of them configured with an IP address. Shifting the elements which don't have address configured to the tail of list sounds fair to me. Thanks.

To keep things simple, we could iterate over the verbs_devs first, and then search for fi_info with same domain name. If there's a match, the fi_info can be pulled out of the list to (create a new list/append to the new list), then add the address info and so on. The remaining fi_info in the original list would be the ones which don't have addresses and they can be appended to the newly created list.